### ISSUE TITLE
`meson -Db_sanitize=address` does not check that `gcc -fsanitize=address` works

### ISSUE THREAD

**Issue by evverx** [mesonbuild/meson#1582]
```sh
$ meson --version
0.40.0.dev1

$ cat main.c
#include<stdio.h>

int main(int argc, char **argv) {
  printf("Hello there.\n");
  return 0;
}

$ cat meson.build
project('tutorial', 'c')
executable('demo', 'main.c')

$ meson -Db_sanitize=address build && ninja-build -C build
The Meson build system
Version: 0.40.0.dev1
Source dir: /home/vagrant/simple
Build dir: /home/vagrant/simple/build
Build type: native build
Project name: tutorial
Native c compiler: cc (gcc 6.3.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Build targets in project: 1
ninja: Entering directory `build'
[2/2] Linking target demo
FAILED: demo
cc  -o demo 'demo@exe/main.c.o' '-fsanitize=address' '-Wl,--no-undefined' '-Wl,--as-needed'
/usr/bin/ld: cannot find /usr/lib64/libasan.so.3.0.0
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by jpakkane** [comment]
It fails in linking, not compiling. This would indicate that GCC does support asan, but the support library for it is not installed.

Under which circumstances would a GCC or Clang not ship with the sanitizers? I don't think I have seen one in years. Is this an issue for self-built compilers or old RHELs?

**Comment by ignatenkobrain** [comment]
@jpakkane it's like that in Fedora as well. you have to manually install libasan

**Comment by evverx** [comment]
> It fails in linking, not compiling. This would indicate that GCC does support asan, but the support library for it is not installed

You are right. I think the "supports" is the wrong word. Probably I should rename it to "works".

> Under which circumstances would a GCC or Clang not ship with the sanitizers?

`Fedora`, `CentOS`, `Ubuntu` and `Debian` ship `libasan` as a standalone package which may not be installed.

**Comment by evverx** [comment]
Sorry, I forgot to mention that `libtsan`, `libubsan` and the other sanitizers are separate packages too.

**Comment by QuLogic** [comment]
Similarly, `-Db_sanitize=memory` 'works', but gcc 6.3.1 says it's an unrecognized argument when compiling.

**Comment by dcbaker** [comment]
it turns out that there are cases where asan isn't available, like less common architectures:
https://673562.bugs.gentoo.org/attachment.cgi?id=558340

**Comment by bonzini** [comment]
My idea was:

* Name it b_sanitizers which is better anyway than b_sanitize
* Do not allow setting b_sanitize anymore on the command line to simplify the implementation
* Make get_option('b_sanitize') return the join of the elements of get_option('b_sanitizers')
